### PR TITLE
:sparkles: Add evidential classification

### PR DESCRIPTION
**Pull request by xuanlongORZ**
- [x] Handle the number of epochs for the simulated annealing
- [x] Add the method to the ReadMe

### CODE REVIEW

File: auto_tutorials_source/tutorial_dec_cubic.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+"""
+Deep Evidential Classification on a Toy Example
+===========================================
```
Comment (o-laurent) [inline auto_tutorials_source/tutorial_dec_cubic.py:6]: Finish the line

File: docs/source/references.rst [hunk 2]
```diff
@@ -8,6 +8,17 @@ Uncertainty Models
 
 The following uncertainty models are implemented.
 
+Deep Evidential Classification
```
Comment (o-laurent) [inline docs/source/references.rst:11]: Add to the ReadMe too

File: torch_uncertainty/losses.py [hunk 3]
```diff
@@ -173,3 +173,198 @@ def forward(
...
+        evidence: Tensor,
+        targets: Tensor,
+    ) -> Tensor:
+        evidence = torch.relu(evidence)
+        alpha = evidence + 1.0
+
+        kl_alpha = (alpha - 1) * (1 - targets) + 1
+
+        ones = torch.ones(
+            [1, self.num_classes], dtype=torch.float32, device=evidence.device
```
Comment (o-laurent) [inline torch_uncertainty/losses.py:292]: the dtype and number of classes should be automatically determined

File: auto_tutorials_source/tutorial_dec_cubic.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+- the classification training routine in the torch_uncertainty.training.classification module
+- the evidential objective: the DECLoss, which lies in the torch_uncertainty.losses file
+- the datamodule that handles dataloaders: MNISTDataModule, which lies in the torch_uncertainty.datamodule
+- the optimizer wrapper in the torch_uncertainty.optimization_procedures module.
+"""
+
+from torch_uncertainty import cli_main, init_args
+from torch_uncertainty.datamodules import MNISTDataModule
+from torch_uncertainty.losses import DECLoss
+from torch_uncertainty.baselines import ResNet
```
[the remark this answers is not in the document]
Comment (o-laurent) [reply auto_tutorials_source/tutorial_dec_cubic.py:33]: This is fully my fault, fixing the other tutorials

[the remark this answers is not in the document]
Comment (xuanlongORZ) [reply auto_tutorials_source/tutorial_dec_cubic.py:33]: you mean we don't need ResNet here ?

File: auto_tutorials_source/tutorial_dec_cubic.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+with ArgvContext(
+    "file.py",
+    "--max_epochs",
+    "25",
+    "--enable_progress_bar",
+    "True",
+    "--version",
+    "vanilla",
+):
+    args = init_args(network=ResNet, datamodule=MNISTDataModule)
```
[the remark this answers is not in the document]
Comment (xuanlongORZ) [reply auto_tutorials_source/tutorial_dec_cubic.py:90]: I didn't find LeNet in baselines

File: torch_uncertainty/losses.py (hunk 3 above)
Comment (o-laurent) [reply torch_uncertainty/losses.py:292]: Yes, you are right! Thanks

File: auto_tutorials_source/tutorial_dec_cubic.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (o-laurent) [reply auto_tutorials_source/tutorial_dec_cubic.py]: Yes, we could add it, but here, we don't really need specific arguments for the LeNet. But you are right that it is not explained clearly!

File: auto_tutorials_source/tutorial_dec_cubic.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (o-laurent) [reply auto_tutorials_source/tutorial_dec_cubic.py]: It is imported to get its particular arguments, but as I wrote in the following comment, they are useless.

File: auto_tutorials_source/tutorial_evidential_classification.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+    entropy = -1 * torch.sum(probs * torch.log(probs), dim=1, keepdim=True)
+    for j in range(4):
+        predicted = torch.argmax(probs[j, :])
+        print(
+            f"Predicted digits for the image {j}: {predicted} with strength "
+            f"{strength[j,0]:.3} and entropy {entropy[j,0]:.3}."
+        )
+
+    # rotate the images by 45 degrees
+    rotated_images = F.rotate(images, 45)
```
Comment (o-laurent) [inline auto_tutorials_source/tutorial_evidential_classification.py:171]: @alafage @xuanlongORZ what do you think of adding RotatedMNIST to the datasets?

File: torch_uncertainty/losses.py (hunk 3 above)
Comment (xuanlongORZ) [reply torch_uncertainty/losses.py:292]: Since it is also used in the one hot function, I think leave it in the class is better

File: torch_uncertainty/losses.py [hunk 7]
```diff
@@ -173,3 +173,198 @@ def forward(
...
+        loss_reg = self._kldiv_reg(evidence, targets)
+
+        loss = loss_dirichlet + annealing_coef * loss_reg
+
+        if self.reduction == "mean":
+            return loss.mean()
+        elif self.reduction == "sum":
+            return loss.sum()
+        else:
+            return loss
```
Comment (o-laurent) [inline torch_uncertainty/losses.py:370]: @alafage, time to split losses.py in a folder?

File: auto_tutorials_source/tutorial_evidential_classification.py (hunk 6 above)
Comment (o-laurent) [reply auto_tutorials_source/tutorial_evidential_classification.py:171]: Crystal clear, thanks!

File: torch_uncertainty/losses.py (hunk 3 above)
Comment (o-laurent) [reply torch_uncertainty/losses.py:292]: Yes, of course, I just meant that the users will have to use partial(DECLoss, num_classes=...) which is something that I would have preferred to avoid...

File: torch_uncertainty/losses.py (hunk 7 above)
Comment (o-laurent) [reply torch_uncertainty/losses.py:370]: Let's keep it this way one last time.

**Comment by o-laurent** [comment]
It seems ready to merge for me. We may need to detail the loss more in the tutorial + add some more tests, but we will do this in the PR on main or just on dev!
Thanks again for your help :)
